src/test/java/uk/gov/ons/ctp/integration/envoycuc/mockclient/MockClient.java

```diff
@@ -0,0 +1,292 @@
+        new RateLimiterClientRequest(domain, product, caseType, ipAddress, uprn, telNo);
+    final IsValidRequest isValidRequest = postRequest(rateLimiterClientRequest);
+
+    if (!isValidRequest.isValid()) { // invalid request - throw a 429
+      final StringBuilder reason = new StringBuilder("Too Many Requests - ");
+      isValidRequest
+          .getLimitStatusList()
+          .forEach(stat -> reason.append(stat.toString()).append(" - "));
+      throw new ResponseStatusException(
+          org.springframework.http.HttpStatus.TOO_MANY_REQUESTS, reason.toString());
+    }
+
+    RateLimitResponse response = new RateLimitResponse();
+    response.setOverallCode("200");
+    response.setStatuses(isValidRequest.getLimitStatusList());
+    final StringBuilder reason = new StringBuilder("Request Successful - ");
+    isValidRequest
+        .getLimitStatusList()
+        .forEach(stat -> reason.append(stat.toString()).append(" - "));
+    log.info(reason.toString());
+    return response;
+  }
+
+  public IsValidRequest postRequest(final RateLimiterClientRequest rateLimiterClientRequest)
+      throws ResponseStatusException {
+
+    List<String> requestKeyList = getKeys(rateLimiterClientRequest);
+    final IsValidRequest isValidRequest =
+        isValidateRequest(requestKeyList, rateLimiterClientRequest);
+    postRequest(
+        requestKeyList,
+        rateLimiterClientRequest); // always post - it burns allowances every time for all scenarios
+
+    return isValidRequest;
+  }
+
+  private List<String> getKeys(RateLimiterClientRequest request) {
+    List<String> keyList = new ArrayList<>();
+    StringBuilder keyBuff =
+        new StringBuilder(request.getProduct().getProductGroup().name().toUpperCase())
+            .append("-")
+            .append(request.getProduct().getIndividual().toString().toUpperCase())
+            .append("-")
+            .append(request.getProduct().getDeliveryChannel().name().toUpperCase())
+            .append("-")
+            .append(request.getCaseType().name().toUpperCase())
+            .append("-");
+    if (request.getUprn() != null && request.getUprn().getValue() != 0) {
+      keyList.add(keyBuff.toString() + "UPRN");
+    }
+    if (request.getIpAddress() != null) {
+      keyList.add(keyBuff.toString() + "IP");
+    }
+    if (request.getTelNo() != null) {
+      keyList.add(keyBuff.toString() + "TELNO");
+    }
+    return keyList;
+  }
+
+  private IsValidRequest isValidateRequest(
+      List<String> requestKeyList, RateLimiterClientRequest f) {
+    final IsValidRequest isValidRequest = new IsValidRequest();
+    for (String requestKey : requestKeyList) {
+      if (!allowanceMap.containsKey(requestKey)) {
+        continue; // not in scope so is valid....
+      }
+      final int noRequestAllowed = allowanceMap.get(requestKey);
+      Map<String, List<Integer>> postedMap = postingsTimeMap.get(requestKey);
+      if (postedMap == null) {
+        return isValidRequest;
```

Comment:
do we want to throw an exception here? since we have slipped through  finding an allowance, so if we can't find a tracking map , I imagine this is an error in the code?

Reply:
The question is - what does the real limiter do?